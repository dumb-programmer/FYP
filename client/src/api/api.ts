export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
