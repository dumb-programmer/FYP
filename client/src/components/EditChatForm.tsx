import { useState } from "react"

export default function EditChatForm({ chat, refetch, onCancel }) {
    const [name, setName] = useState(chat.name);

    const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const response = await fetch(`http://localhost:3000/chats/${chat._id}`, {
            method: "PATCH",
            body: JSON.stringify({ name }),
            headers: {
                "Content-Type": "application/json"
            },
            credentials: "include",
            mode: "cors"
        });
        if (response.ok) {
            onCancel();
            refetch();
        }
    }

    return <form className="flex flex-col gap-2" onSubmit={handleSubmit}>
        <input className="input input-bordered" type="text" value={name} onChange={(e) => setName(e.target.value)} />
        <div className="flex justify-end gap-2">
            <button className="btn" onClick={onCancel}>Cancel</button>
            <button className="btn btn-primary" type="submit">Submit</button>
        </div>
    </form>;
}