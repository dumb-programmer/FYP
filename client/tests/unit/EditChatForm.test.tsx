import React from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi } from "vitest";
import EditChatForm from "../../src/components/EditChatForm";
import userEvent from "@testing-library/user-event";
import { updateChat } from "../../src/api/api";


vi.mock("../../src/api/api.ts", async () => ({ ...await vi.importActual("../../src/api/api.ts"), updateChat: vi.fn().mockReturnValue({ ok: true }) }))

const chat = { _id: crypto.randomUUID(), name: "test" }
const mockRefetch = vi.fn();
const mockOnCancel = vi.fn();
const renderComponent = () => render(<EditChatForm chat={chat} refetch={mockRefetch} onCancel={mockOnCancel} />);

describe("EditChatForm", () => {
    it("renders correctly", () => {
        renderComponent();

        const component = screen.getByTestId("edit-chat-form");

        expect(component).toMatchSnapshot();
    });

    it("input is populated with chat name by default", () => {
        renderComponent();

        const input = screen.getByRole("textbox");

        expect(input.value).toEqual(chat.name);
    });

    it("input works", async () => {
        renderComponent();

        const input = screen.getByRole("textbox");

        await userEvent.clear(input);
        await userEvent.type(input, "new chat");

        expect(input.value).toEqual("new chat");
    });

    it("onCancel is called on cancel button click", async () => {
        renderComponent();

        const cancelBtn = screen.getAllByRole("button")[0];

        await userEvent.click(cancelBtn);

        expect(mockOnCancel).toHaveBeenCalled();
    });

    it("chat is updated on submit", async () => {
        renderComponent();

        const input = screen.getByRole("textbox");

        await userEvent.clear(input);
        await userEvent.type(input, "new chat");

        const submitBtn = screen.getAllByRole("button")[1];

        await userEvent.click(submitBtn);

        expect(updateChat).toHaveBeenCalled();
        expect(updateChat).toHaveBeenCalledWith(chat._id, "new chat");

        expect(mockRefetch).toHaveBeenCalled();
    });
})