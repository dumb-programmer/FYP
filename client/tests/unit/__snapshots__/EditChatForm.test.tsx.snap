// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditChatForm > renders correctly 1`] = `
<form
  class="flex flex-col gap-2"
  data-testid="edit-chat-form"
>
  <input
    class="input input-bordered"
    type="text"
    value="test"
  />
  <div
    class="flex justify-end gap-2"
  >
    <button
      class="btn btn-sm text-xs"
      type="button"
    >
      Cancel
    </button>
    <button
      class="btn btn-primary btn-sm text-xs"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
